### ISSUE TITLE
expose a simple friendlier GET API for retrieving contract state, eg. for use with curl.

### ISSUE THREAD

**Issue by gogo2464** [freenet/freenet-core#904]
I would like to connect mobile and web apps to freenet. Could we create a way to interact directly to contracts, delegate and user interface directly from web or apps please?

**Comment by sanity** [comment]
We could also support puts using HTTP PUT, although the user would need to provide the contract WASM in addition to the updated state or delta.

**Comment by iduartgomez** [comment]
A couple comments/caveats first.

There are some things are not keen to using regular HTTP requests like the subscribe and updates since the way it works it would require polling from the client side so that's gonna feel a bit awkward compared to the more natural websocket approach where we are essentially implementing a pub/sub model though it.

That said, I can see the value of this and ease of use if you want to do approach managing contracts in a more traditional style, like through a REST API.

The implementation should be done via adding new path handlers to the existing http gateway server handling the websocket connections. See `crates/core/src/server/http_gateway.rs:as_router` function. 

We would add new route handlers under the `/contract/api/<OPERATION>` (so for example `/contract/api/get`). After that we can add support in the stdlib to support the API, since we will re-using the types defined through the common schemas and serializing/deserializing flatbuffers, need to make that friendlier to the user.

But we can break the task in 2 separate PR's so is easier to implement and review. So first do the necessary changes to the server. The server needs to handle the requests to the contract executor and then wait for an answer, since all that is running on it's own task/thread, you can check how this is done in the websocket implementation at `crates/core/src/client_events/websocket.rs:websocket_interface` .

The regular HTTP request one should be a simplier version of this, and we would need a separate handler for each operation (let's just start with get since that's relatively simple). I could get this done quickly but I will leave it open to implement to anybody who wants to help contributing so they get more familiar with how the whole application is setup.

P.S. Also must comment that curl can upgrade connections to use websockets, although it ain't the more natural/easiest to use thing, since this is what this initially spawned the request.

**Comment by iduartgomez** [comment]
If you need any help while implementing feel free to just post them here and I'll help you moving as fast as possible.

**Comment by gogo2464** [comment]
@iduartgomez yes I need help. could we talk on freenet server please?

**Comment by gogo2464** [comment]
I have some ideas about how to expose contract state. I am also working on how to expose an API filtered by user inetrface.

**Comment by Julian-Dumitrascu** [comment]
> I have some ideas about how to expose contract state. I am also working on how to expose an API filtered by user inetrface.

Would you enjoy a public talk or a private talk about this? Either way, we can schedule a talk soon.

**Comment by gogo2464** [comment]
> > I have some ideas about how to expose contract state. I am also working on how to expose an API filtered by user inetrface.
> 
> Would you enjoy a public talk or a private talk about this? Either way, we can schedule a talk soon.

Yes I would love to enjoy a private talk. I also think a public dev documentation could benefit to the dev community. I just would not like to be recorded.

**Comment by Julian-Dumitrascu** [comment]
> recorded

I've started [this conversation](https://github.com/Sol-Global-Management/data-management/issues/10) to help anyone make such requirements.

**Comment by sanity** [comment]
🤖 **Relabeled for consistency**

Applied labels:
- `T-feature`
- `P-medium`
- `E-medium`
- `A-contracts`
- `A-developer-xp`
- `S-needs-design`

**Reasoning:** The report requests a new, simpler GET API to retrieve contract state for use by web/mobile (curl, apps), which is a new functionality rather than a bug or pure docs change, so T-feature fits. It's not blocking or urgent, so P-medium is appropriate. Implementing a secure, simple REST-like endpoint for contract state likely requires moderate design and implementation work (API design, auth, contract access), so E-medium. The primary area is contracts (exposing contract state) and it also impacts developer experience for app authors. The issue requires API design decisions before implementation, so S-needs-design is included.

Previous labels: `good-first-issue`

**Comment by sanity** [comment]
# Issue Priority Assessment

Generated: 2026-02-22T14:50:29.223161528+00:00
Issue hash: `5b7ed713`

## Value Assessment

Serving contract state over a simple HTTP GET would have concrete value for the subset of mobile and web developers who’d rather poke a REST endpoint (or use `curl`) than manage the existing websocket lifecycle, which currently funnels them through the contract UI handler that unpacks WASM state to a temp web bundle (see `path_handlers::contract_home` for the workflow that connects to `ClientConnection`, issues a `ContractRequest::Get`, waits on the node reply, and then streams the unpacked UI; `crates/core/src/server/path_handlers.rs:22-187`). Right now those clients either need to implement a full websocket client (or a clunky websocket upgrade from `curl`), so their experience is more cumbersome than the “simple GET” the issue asks for.

The feature is well-aligned with the HTTP gateway module’s architecture: `HttpClientApi::as_router_with_attested_contracts` already merges the v1/v2 Axum routers and shares the pipes into the node via `ClientConnection` and attested tokens (`crates/core/src/server/client_api.rs:69-117`). Adding `/contract/api/get` handlers there would reuse those same request/response channels that the websocket path already leverages in `websocket_interface` (`crates/core/src/client_events/websocket.rs:529-578`), so implementation effort would largely be wiring the existing executor call/response into a synchronous HTTP handler rather than inventing a new pipeline. That keeps the REST endpoint on-strategy with the project’s goal of broad, developer-friendly contract access without duplicating contract logic.

This issue is not a blocker for other work because it’s purely a developer-experience enhancement; nothing downstream depends on having this simpler GET API in place before moving forward. It’s also tagged “needs design,” which suggests it’s intended to be queued until a contributor is ready to define the API surface, so we can schedule it whenever capacity allows without stalling core progress.

There’s no pressing time constraint—no release or SEC deadline is cited—so the priority stays medium as labeled. The value is real for API-focused integrators, but it can wait for the usual sprint cadence once there’s clarity on the exact URL/serialization choices.

## Cost Assessment

The HTTP gateway is already wired through `HttpClientApi::as_router_with_attested_contracts`, which merges the V1 and V2 routers and wires in the shared request channel plus attested-token state (`crates/core/src/server/client_api.rs:79`). Both version routers simply mount `/v*/contract/web/...` for contract web UIs (`crates/core/src/server/client_api/v1.rs:6`, `crates/core/src/server/client_api/v2.rs:9`), so exposing `/contract/api/get` will require touching those modules to register new handlers while leaving the shared `HttpClientApiRequest`/`attested_contracts` plumbing intact.

The handler itself will likely mirror `contract_home`, since that routine already constructs a new `ClientConnection::NewConnection`, waits for the assigned client ID, and then issues a `ContractRequest::Get` before processing the `HostCallbackResult` and serializing the response (`crates/core/src/server/path_handlers.rs:41`). Any simple GET API must therefore reuse that request/response choreography, but instead of unpacking a web bundle it will just unwrap the `HostResponse::ContractResponse` and format the state; doing so safely requires examining how the websocket side does it in `websocket_interface`, because that logic also starts a client connection and routes `ClientRequest::ContractOp` through the core executor (`crates/core/src/client_events/websocket.rs:529`, `crates/core/src/server/mod.rs:200`).

Because all contract APIs flow through the same executor loop, adding a synchronous HTTP route carries the typical risks of deadlocking the executor if the new handler waits on the node thread in the wrong place, and token cleanup still assumes clients get removed via `ClientRequest::Disconnect` (`crates/core/src/server/mod.rs:219`). Tests therefore should include an integration hit against the HTTP router to confirm a GET request actually goes through `ClientRequest::ContractOp`, that tokens dropped when the handler finishes, and that the response is serialized in a client-friendly form. Coverage will need to exercise both V1 and V2 routes plus the common path handler to catch regressions, so expect a modest integration harness around the existing `serve_client_api_in_impl` setup.

---
*Generated by issue-prioritizer*

**Comment by sanity** [comment]
🤖 **Automated ranking by Torvan.** Not a maintainer's assessment of this issue.

| Measure | Score |
|---|---:|
| Value | 0.552 |
| Cost | 0.716 |
| Priority | 353 of 593 |

Value and cost run 0 to 1 across the 593 open issues in the freenet repositories: value 1.00 is the most valuable, cost 0.00 the cheapest. Issues are compared against each other rather than scored in isolation. The queue alternates between the most valuable work and the best value for the effort, so neither expensive-but-important nor cheap-and-quick work crowds the other out.

Ranked 2026-08-31, and liable to change. Adding information here, whether that is scope, a reproduction, or a correction to this assessment, causes the issue to be reassessed and this comment updated.

**Comment by gogo2464** [comment]
`fdev execute` currently supports only local contracts

**Comment by gogo2464** [comment]
`crates/core/src/server/http_gateway.rs:as_router` has been replaced with `crates/core/src/server/client_api.rs` .
